### PR TITLE
test_containers_smartstate_analysis + Framework changes

### PR DESCRIPTION
**Pull request by gshefer**
Purpose or Intent
=================

- added:
     cfme/tests/containers/test_containers_smartstate_analysis.py
     Smart State analysis functionality check for single container image.
- added important functions to framework.

{{pytest: cfme/tests/containers/test_containers_properties.py -v --use-provider container }}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit f9480192e83bce2e432bf583d6811d675698bcaf:

`cfme/tests/containers/test_containers_smartstate_analysis.py`:
- :heavy_exclamation_mark: Line 77:1: **W391** *blank line at end of file*

Please, rectify these issues :smirk: .

Commit inspection report:

**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a03bef06356286c12a3b1db23aaa5772c50f9a66:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

File: utils/ssh.py [hunk 1]
```diff
@@ -480,6 +480,14 @@ def set_initial_file_end(self):
             fstat = sshtail._sftp_client.stat(self._remote_filename)
             self._remote_file_size = fstat.st_size  # Seed initial size of file
 
+    def raw_output(self):
+        """Returns the output of the log as one (raw) string"""
+        return '\n'.join(self.lines_as_list())
```
Comment (jaryn) [inline utils/ssh.py]: 1. You don't need to create a list for this. Just use the generator object:
    '\n'.join(line for line in self)
2. I don't like calling this raw_output as it is not raw. The newlines are gonna be mangled depending on the platform we are running on. The '\n' is <CR> on Mac, <LF> on unix and <CR><LF> on windows

**Comment by dajoRH** [comment]
Lint report for commit 2d93597440737be8925412e9308676d7e382b3a7:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

File: utils/ssh.py (hunk 1 above)
Comment (gshefer) [inline utils/ssh.py]: 1. Look at lines_as_list it did exactly [line for line in self] -> this is why I created this.
2. any suggestion for name and generic '\n'?

Comment (jaryn) [inline utils/ssh.py]: 1. No it is not exactly the same.

        (line for line in something)

 is a generator

      [line for line in something]

 is a list

Comment (jaryn) [inline utils/ssh.py]: I would do it like this:

     def __iter__(self):
          for line in self.raw_lines:
               yield line.rstrip()

     def raw_lines(self):
         with self as sshtail:
             fstat = sshtail._sftp_client.stat(self._remote_filename)
             if self._remote_file_size is not None:
                 if self._remote_file_size < fstat.st_size:
                     remote_file = self._sftp_client.open(self._remote_filename, 'r')
                     remote_file.seek(self._remote_file_size, 0)
                     while (remote_file.tell() < fstat.st_size):
                         line = remote_file.readline()                                # Note the  missing rstrip() here!
                         yield line
             self._remote_file_size = fstat.st_size

     def raw_string():
          retrun ''.join(l for l in raw_lines())

**Comment by dajoRH** [comment]
Lint report for commit f9807c366e0c652b5ae20d7ef93f9f9cb7c1f8cf:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a2c46b798d433dcb84d2bc6633aac30be51f62c7:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

File: cfme/tests/containers/test_containers_smartstate_analysis.py [hunk 2]
```diff
@@ -28,6 +28,14 @@
                         'OpenSCAP Results': len, 'OpenSCAP HTML': len, 'Last scan': len}
 
 
+def wait_for_all_tasks_finish():
+    # wait until all tasks are finished
+    sel.force_navigate('tasks_all_vm')
+    if len(tasks.all_vm_analysis_tasks()):
```
Comment (mfalesni) [inline cfme/tests/containers/test_containers_smartstate_analysis.py]: Nonempty list is als otrue, len() is redundant here.

Comment (gshefer) [inline cfme/tests/containers/test_containers_smartstate_analysis.py]: modified accordingly...

File: cfme/configure/tasks.py [hunk 3]
```diff
@@ -173,6 +174,11 @@ def all_other_tasks(**filter_kwargs):
     return _get_tasks("tasks_all_other", **filter_kwargs)
 
 
+def delete_all_tasks():
+    toolbar.select('Delete Tasks', 'Delete All', invokes_alert=True)
```
Comment (jkrocil) [inline cfme/configure/tasks.py]: This needs to include navigation if we are not yet on the right page (`force_navigate`).

File: cfme/configure/tasks.py [hunk 4]
```diff
@@ -252,3 +257,9 @@ def is_analysis_finished(name, task_type='vm', clear_tasks_after_success=True):
                             task_name=tabs_data['task'].format(name),
                             expected_status=tabs_data['state'],
                             clear_tasks_after_success=clear_tasks_after_success)
+
+
+def wait_for_analysis_finish(task_name, delay, timeout):
```
Comment (jkrocil) [inline cfme/configure/tasks.py]: This function must support all types of analyses (`vm, host, datastore, cluster`) and not just `vm`. In addition, the `delay` and `timeout` parameters must have reasonable default values.

File: cfme/tests/containers/test_containers_smartstate_analysis.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    pytest.mark.meta(server_roles='+smartproxy'),
+    pytest.mark.uncollectif(
+        lambda: current_version() < "5.6"),
+    pytest.mark.usefixtures('setup_provider'),
+    pytest.mark.tier(1)]
+pytest_generate_tests = testgen.generate(
+    testgen.container_providers, scope='function')
+
+list_tbl = CheckboxTable(table_locator="//div[@id='list_grid']//table")
+tasks_tbl = CheckboxTable(table_locator="//table[@class='table table-striped" +
```
Comment (jkrocil) [inline cfme/tests/containers/test_containers_smartstate_analysis.py]: The `+` operator here is not necessary; python will concatenate the strings even without it, as long as the whole string is wrapped in parentheses.

File: cfme/configure/tasks.py (hunk 4 above)
Comment (jkrocil) [inline cfme/configure/tasks.py]: Also, this should be called `wait_analysis_finished`, you can check out naming of functions above (`is_analysis_finished`, `is_task_finished`, ...)

**Comment by dajoRH** [comment]
Lint report for commit b225f3fd4ab057995e9cc49ddd6d1aab2044229f:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 3a485117bdc37b6c87946fef6288f1ca3cb3b25c:

Everything seems all right from lint perspective :smile: .

Commit inspection report:
* 3a485117bdc37b6c87946fef6288f1ca3cb3b25c
  * Merge commit detected!!! :bangbang:

**The PR might not be acceptable in this state** :rage4:.
There is a merge commit in the branch which can cause some problems during squashing. Please take **EXTRA care** when squashing!
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Pull request closed without merging**